File: e19.pl
my $day = 1;
my $sundays = 0;
for my $y (0..100){
    for my $mo (0..11){
        print "$day ";
        if ($day == 0 && $y != 0) {
            $sundays += 1;
        }
        if ($mo == 0 || $mo == 2 || $mo == 4 || $mo == 6 || $mo == 7 || $mo == 9 || $mo == 11) {
            $day = ($day + 3) % 7;
        } elsif ($mo == 1) {
          if ($y % 4 == 0 && $y != 0) {
            $day = ($day + 1) % 7;
          }
        } else {
          $day = ($day + 2) % 7;
        }
    }
    print "\n"
}
print "Number of Sundays is $sundays\n";

File: e19.fs

let rec answer day sundays mo y =
    if y = 101  then sundays else
    if mo = 12  then (answer day sundays 0 (y + 1)) else
        let newsun = (if (day = 0) && not (y = 0) then sundays + 1 else sundays)
        if (mo = 0 || mo = 2 || mo = 4 || mo = 6 || mo = 7 || mo = 9 || mo = 11) then
            (answer ((day + 3) % 7) newsun (mo + 1) y) else
        if (mo = 1) then
            (answer (if ((y % 4 = 0) && not (y = 0)) then ((day + 1) % 7) else day) newsun (mo + 1) y)
        else
            (answer ((day + 2) % 7) newsun (mo + 1) y)

printfn "%d" (answer 1 0 0 0)
